src/dist/http.rs

```diff
@@ -306,82 +315,67 @@ mod server {
-            .set_issuer_name(&name)
-            .context("failed to set issuer name")?;
-
-        // Add the SubjectAlternativeName
-        let extension = openssl::x509::extension::SubjectAlternativeName::new()
-            .ip(&addr.ip().to_string())
-            .build(&builder.x509v3_context(None, None))
-            .context("failed to build SAN extension for x509")?;
-        builder
-            .append_extension(extension)
-            .context("failed to append SAN extension for x509")?;
-
-        // Add ExtendedKeyUsage
-        let ext_key_usage = openssl::x509::extension::ExtendedKeyUsage::new()
-            .server_auth()
-            .build()
-            .context("failed to build EKU extension for x509")?;
-        builder
-            .append_extension(ext_key_usage)
-            .context("fails to append EKU extension for x509")?;
-
-        // Finish the certificate
-        builder
-            .sign(&privkey, openssl::hash::MessageDigest::sha1())
-            .context("failed to sign x509 with sha1")?;
-        let cert: openssl::x509::X509 = builder.build();
-        let cert_pem = cert.to_pem().context("failed to create pem from x509")?;
-        let cert_digest = cert
-            .digest(openssl::hash::MessageDigest::sha256())
-            .context("failed to create digest of x509 certificate")?
-            .as_ref()
-            .to_owned();
-
+        let mut rng = OsRng;
+        let bits = 2048;
+        let rsa_key = rsa::RsaPrivateKey::new(&mut rng, bits)?;
+
+        let line_ending = rsa::pkcs8::LineEnding::CRLF;
+        let sk_pkcs8 = rsa::pkcs8::EncodePrivateKey::to_pkcs8_pem(&rsa_key, line_ending)?;
+        let pk_pkcs8 = rsa::pkcs8::EncodePublicKey::to_public_key_pem(&*rsa_key, line_ending)?;
+
+        // convert to picky
+        let sk = PrivateKey::from_pem_str(sk_pkcs8.as_str())?;
+        let pk = PublicKey::from_pem_str(pk_pkcs8.as_str())?;
+        let today = chrono::Utc::now().naive_utc();
+        let expires = today + chrono::Duration::days(365);
+        let start = UTCDate::new(
+            today.year() as u16,
+            today.month() as u8,
+            today.day() as u8,
+            today.time().hour() as u8,
+            today.time().minute() as u8,
+            today.time().second() as u8,
+        )
+        .unwrap();
+        let end = UTCDate::new(
+            expires.year() as u16,
+            expires.month() as u8,
+            expires.day() as u8,
+            expires.time().hour() as u8,
+            expires.time().minute() as u8,
+            expires.time().second() as u8,
+        )
+        .unwrap();
+        let extended_key_usage = ExtendedKeyUsage::new(vec![picky::oids::kp_server_auth()]);
+        let name = addr.to_string();
+        let issuer_name = DirectoryName::new_common_name(name.clone());
+        let subject_name = DirectoryName::new_common_name(name);
+        let octets = match addr.ip() {
+            IpAddr::V4(inner) => inner.octets().to_vec(),
```

Comment:
I'm guessing the usage of `octets()` is wrong.

From [RFC](https://datatracker.ietf.org/doc/html/rfc5280#appendix-A.2):

> The syntax of iPAddress MUST be as described in [Section 4.2.1.6](https://datatracker.ietf.org/doc/html/rfc5280#section-4.2.1.6) with
   the following additions specifically for name constraints.  For IPv4
   addresses, the iPAddress field of GeneralName MUST contain eight (8)
   octets, encoded in the style of [RFC 4632](https://datatracker.ietf.org/doc/html/rfc4632) (CIDR) to represent an
   address range [[RFC4632](https://datatracker.ietf.org/doc/html/rfc4632)].  For IPv6 addresses, the iPAddress field
   MUST contain 32 octets similarly encoded.  For example, a name
   constraint for "class C" subnet 192.0.2.0 is represented as the
   octets C0 00 02 00 FF FF FF 00, representing the CIDR notation
   192.0.2.0/24 (mask 255.255.255.0).


But, `IpAddr`'s `octects` is:

```rust
use std::net::Ipv4Addr;

let addr = Ipv4Addr::new(127, 0, 0, 1);
assert_eq!(addr.octets(), [127, 0, 0, 1]);
```

Reply:
Thanks, I will have a look at this.

Reply:
@Xuanwo I've fixed this issue by manually inserting the mask, however the issue still persists:

```
[2023-05-10T13:04:33Z ERROR rustls::conn] TLS alert received: AlertMessagePayload {
        level: Fatal,
        description: BadCertificate,
    }
```

Since this also happens in my other PR, which only disables the `reqwest/default-tls` feature, I guess this is a bug of reqwest or rustls.

Reply:
> Since this also happens in my other PR, which only disables the `reqwest/default-tls` feature, I guess this is a bug of reqwest or rustls.

Ok, maybe it's ok for us to enable `reqwest/default-tls` in test?

Reply:
> Ok, maybe it's ok for us to enable `reqwest/default-tls` in test?

I can do that, but this seems like to be a bug of `rustls`.
Perhaps this could even happen in real-world, though I'm not entirely sure.

@Xuanwo Does the CI for `sccache-dist` uses self-signed certificate generated by `sccache` or use a pre-generated certificate?

Reply:
> Does the CI for `sccache-dist` uses self-signed certificate generated by `sccache` or use a pre-generated certificate?

I assume that `sccache-dist` will create a new self-signed certificate when the scheduler is launched.

Reply:
Hi! Sorry for the intrusion, but I'm wondering if this isn't related to rustls/rustls#184? `reqwest` is still on `rustls 0.20.0`, and [support for peers named with IP addresses](https://github.com/rustls/rustls/tree/main#release-history) was added in `rustls 0.21.0`

The Cargo.lock's [rustls is still on 0.20.7](https://github.com/NobodyXu/sccache/blob/dist-server-rewrite/Cargo.lock#L2290-L2300), which is probably making the CI tests fail?

Reply:
@saiintbrisson Thanks, I think you are right!

I just discovered the same thing when looking at the reqwest PR https://github.com/seanmonstar/reqwest/pull/1791#issuecomment-1485813835 for updating rustls to v0.21.0

Let's wait for a new release in reqwest, which probably won't take too long.

Reply:
Fresh from the Crates.io factory: https://github.com/seanmonstar/reqwest/releases/tag/v0.11.18

Reply:
I've updated reqwest to v0.11.8 and rustls to v0.21.0, it indeed fixed the bad certificate error, but now we got a new error:

```
[2023-05-17T02:11:22Z ERROR rustls::conn] TLS alert received: AlertMessagePayload {
        level: Fatal,
        description: UnknownCA,
    }
```

@saiintbrisson @cpu Any idea how to fix this error?

Reply:
Could you pull out a smaller reproducer and include the problematic peer certificate? That error is fairly generic and could have many root causes.

Reply:
> include the problematic peer certificate

The peer certificate is self-signed and generated at runtime in [here](https://github.com/mozilla/sccache/pull/1742/files#diff-6d63828d1891ff3303de7191b656bc451d27703548c42bd9786fcd1c157aa459R317).

Reply:
Hi! I've spent some time looking into it and turns out there are still two rustls versions. I believe the side receiving the certificate is still on rustls 0.20.7

```
├── rouille v3.6.2
...
│   ├── tiny_http v0.12.0
│   │   ├── rustls v0.20.7 (*)
│   │   ├── rustls-pemfile v0.2.1
│   │   │   └── base64 v0.13.1
```

The `TLS alert received` message is only present on [0.20.7](https://github.com/rustls/rustls/blob/v/0.20.7/rustls/src/conn.rs#L1313).
0.21.0 does not have [this error message anymore](https://github.com/rustls/rustls/blob/4c973b6b10461808e9207aba64ff6c5c00124547/rustls/src/common_state.rs#L464-L465), it is gone from the codebase, the remaining one is `TLS alert warning received` (notice the "warning").

In fact, even the module was renamed from `conn` to `common_state`, but the logs still show `rustls::conn`.

Reply:
I'm not sure if the different versions are the reason behind the error, it may be completely unrelated problem

Reply:
@saiintbrisson Could it be caused by https://github.com/seanmonstar/reqwest/pull/1835 ?

Reply:
I've also submit a PR to upgrade rustls used in tiny-http https://github.com/tiny-http/tiny-http/pull/242

Reply:
> Could it be caused by [seanmonstar/reqwest#1835](https://github.com/seanmonstar/reqwest/pull/1835) ?

You can run the tests locally and use the reqwest git repository directly on cargo.toml
```
$ cargo test --no-default-features --features=dist-tests test_dist_ -- --test-threads 1
```
You must have docker installed to run these (or podman, requires changing all `docker` command spawns on `tests/harness/mod.rs`)

Reply:
Thanks, I would have a try when I'm free

Reply:
I've run the tests here using the updated reqwest branch and the error persists

Reply:
Ehhh, @saiintbrisson can you also try https://github.com/tiny-http/tiny-http/pull/242 ?

Reply:
This one is a bit tricky as we don't depend on the crate directly

Reply:
> This one is a bit tricky as we don't depend on the crate directly

Yeah, it's a dep brings in by rouille.

Could use patching to replace the dep though

Reply:
Imho we should get rid of roullie as in https://github.com/paritytech/cachepot/pull/131

Reply:
@drahnr I agree, but that PR is quite large and porting it all at once takes a lot of effort.

Reply:
Also good to notice that `trust-dns-resolver` is still on the older version of rustls. I'm guessing this PR might need to wait a bit longer until the ecosystem catches up?

Reply:
Updated trust-dns-resolver v0.22 to latest v0.24 (renamed to hickory-dns) and also updates webpki-roots to v0.25.

Confirmed that rustls v0.21 has been removed, but CI failure persists.

I think it's likely something in the certificate generation that caused the error.